lib/token.cpp

```diff
@@ -53,6 +54,15 @@ Token::~Token()
     delete mImpl;
 }
 
+/*
+* Get a TokenRange which starts at this token and contains every token following it in order
+* e.g. for the sequence of tokens A B C D E, C.toEnd() would yield the Range C D E
+*/
+TokenRange<const Token> Token::toEnd() const
+{
+    return TokenRange<const Token>(this, nullptr);
+}
```

Comment:
What about another method `until`?

```cpp
TokenRange<const Token> Token::until(const Token* end) const
{
    return TokenRange<const Token>(this, end);
}
```

Reply:
I'm concerned that it would be ambiguous as to whether that would be inclusive or exclusive.

I think ```until``` would render toEnd largely redundant as I'd just as soon type ```until(nullptr)``` rather than ```toEnd```.

Reply:
> I'm concerned that it would be ambiguous as to whether that would be inclusive or exclusive.

I would assume it would always be one past the end as its consistent with the standard library, but this might be better as a free function anyways.

Reply:
We could add a `makeTokenRange` function(which should automatically deduce the type):

```cpp
template<class T, REQUIRES("T must be a Token class", std::is_convertible<T*, const Token*>)>
TokenRange<T> makeTokenRange(T* start, T* last)
{
    return TokenRange<T>{start, last};
}
```

Reply:
standard library is usually up to but not including the end right? i.e. the range is [first, last)

I did consider a factory method from which the type could be deduced but I found it didn't save that many characters over just calling the constructor. Now that there is TokenRange and ConstTokenRange to remove the template parameter there doesn't seem much point.

Reply:
> i.e. the range is [first, last)

That is what I understand. You can usually give the end iterator as last.

Reply:
I've decided the usefulness of until outweighs the potential ambiguity which can be mitigated with documentation so went ahead and implemented it.